### PR TITLE
BAU: Check internal and external links

### PR DESCRIPTION
**Pull request by mrwilson**
This PR introduces checking for internal links (within the GDS Way's own content) and external links (content on other sites).

We don't want to have broken links in the GDS Way as this is a bad and confusing user experience. This was highlighted in issue #700.

This PR builds on existing work in https://github.com/alphagov/tech-docs-template/pull/187 and would aim to be folded into the tech docs template eventually.

This PR is also a place for discussion around:
- updated content and new external links prompted by the link checker
- how to handle rate-limiting by GitHub when checking links to repos, of which there are many in the project

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -1,23 +1,23 @@
 GEM
```
Comment (36degrees) [inline Gemfile.lock:1]: There's quite a lot of dependencies being bumped here which seem unrelated to the commit they're in ('ignore GitHub links') – can we split them into a separate commit or PR?

File: source/partials/_nav-version-control-deployments.html.erb [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <ul>
   <li><a href="/standards/source-code/">How to store source code</a></li>
+  <li><a href="/standards/source-code/index.html">How to store source code</a></li>
```
Comment (36degrees) [inline source/partials/_nav-version-control-deployments.html.erb:3]: Isn't this a duplicate of the link above?

File: source/standards/optimise-frontend-perf.html.md.erb [hunk 3]
```diff
@@ -96,7 +96,7 @@ You can find out more about improving your website’s frontend performance by r
 The Service Manual has more suggestions about [how you can test frontend performance][].
 
 [frontend performance]: https://www.gov.uk/service-manual/technology/how-to-test-frontend-performance
-[minification]: https://minifycode.com/
+[minification]: https://web.dev/reduce-network-payloads-using-text-compression/
```
Comment (36degrees) [inline source/standards/optimise-frontend-perf.html.md.erb:99]: We've updated this URL but removed the link from minification on line 21. Is that intentional? Is this link referred to from anywhere else?

File: source/standards/principle-least-access.html.md.erb [hunk 4]
```diff
@@ -43,6 +43,6 @@ If you’re using the gds-users account to log into your AWS accounts, you shoul
 
 - [NCSC - 10 steps to cyber security](https://www.ncsc.gov.uk/collection/10-steps)
 - [NCSC - privileged user management](https://www.ncsc.gov.uk/guidance/introduction-identity-and-access-management#section_6)
-- [NIST Special Publication 800-53 - AC-6 least privilege][polp]
+- [NCSC - preventing lateral movement][polp]
 
-[polp]: https://csrc.nist.gov/Projects/risk-management/sp800-53-controls/release-search#!/control?version=5.1&number=AC-6
```
Comment (36degrees) [inline source/standards/principle-least-access.html.md.erb:48]: I don't have the context here but this link appears to resolve fine. If we're changing the link just because we think the NCSC guidance is better or more appropriate, should we split this change out as it's not broken?

**Review by 36degrees — COMMENTED** [review]
Thanks for picking this up, @mrwilson! A few things I'm not 100% sure about – might just be because I'm missing context.

My Ruby is rusty at best, so you might want a second set of eyes on some of this, depending on your confidence in the changes.

File: Gemfile.lock (hunk 1 above)
Comment (mrwilson) [reply Gemfile.lock:1]: Yep, we can be more specific with this, I "just" copied the working Gemfile.lock in order to get to green,.

File: source/partials/_nav-version-control-deployments.html.erb (hunk 2 above)
Comment (mrwilson) [reply source/partials/_nav-version-control-deployments.html.erb:3]: Yep, will remove

File: source/standards/optimise-frontend-perf.html.md.erb (hunk 3 above)
Comment (mrwilson) [reply source/standards/optimise-frontend-perf.html.md.erb:99]: This was a mistake, I've added the link back

File: source/standards/principle-least-access.html.md.erb (hunk 4 above)
Comment (mrwilson) [reply source/standards/principle-least-access.html.md.erb:48]: Yep, I changed this in an early draft to get it working and didn't change it back, reverted in next change.

**Comment by domoscargin** [comment]
How much do we care about hashes if the target page at least loads?

For instance, we're not catching the fact that this link:
https://github.com/alphagov/gds-way/blob/db16b146f1469d7538497f30845e0a4f5704edf1/source/manuals/programming-languages/html.html.md.erb#L15 has updated to https://github.com/alphagov/govuk-frontend#browser-and-assistive-technology-support

Looks like we could enable `check_external_hash` (and `check_internal_hash`?) to catch these.

**Comment by mrwilson** [comment]
@domoscargin That's a very good shout.

According to the HTMLProofer docs, the default value for both of those flags is true, however I will add them to the configuration for the sake of explicitness and not relying on default behaviour that may change.

**Review by richardTowers — APPROVED** [review]
Looks good to me!

Only thing that I think will be slightly annoying is the inevitable "I just want to fix this typo, but to get the build working I need to fix these unrelated broken links". But that's probably better than having broken links, and I don't think an alternative cadence (e.g. daily check with a slack message, doesn't fail the build) would work given the limited ownership of GDS way.
